### ISSUE TITLE
[API Proposal]: APIs for working with unix file mode

### ISSUE THREAD

**Issue by tmds** [dotnet/runtime#67837]
### Background and motivation

Add direct support for getting/setting Unix file permissions and special bits.

These APIs allow to restrict/extend access to files and directories from .NET.

There have been some requests for these APIs: https://github.com/dotnet/runtime/issues/925, https://github.com/dotnet/runtime/issues/928, https://github.com/dotnet/runtime/issues/17540.

The tar implementation can make use of them (https://github.com/dotnet/runtime/issues/65951) without resorting to internal APIs.

### API Proposal

```C#
namespace System.IO;

[System.FlagsAttribute]
public enum UnixFileMode
{
// From https://github.com/dotnet/runtime/issues/65951.
    None = 0,
    OtherExecute = 1,
    OtherWrite = 2,
    OtherRead = 4,
    GroupExecute = 8,
    GroupWrite = 16,
    GroupRead = 32,
    UserExecute = 64,
    UserWrite = 128,
    UserRead = 256,
    StickyBit = 512,
    SetGroup = 1024, // or, more generic name: GroupSpecial
    SetUser = 2048,  // or, more generic name: UserSpecial

// Maybe:
    AccessPermissionMask = UserRead | UserWrite | UserExecute | GroupRead | GroupWrite | GroupExecute | OtherRead | OtherWrite | OtherExecute,
    DefaultFileOpenPermissions = UserRead | UserWrite | GroupRead | GroupWrite | OtherRead | OtherWrite
}

static partial class Directory
{
    // Set mode when creating file, returns SafeFileHandle.
    public static SafeFileHandle OpenHandle (string path, FileMode mode = FileMode.Open, FileAccess access = FileAccess.Read, FileShare share = FileShare.Read, FileOptions options = FileOptions.None, long preallocationSize = 0, UnixFileMode unixCreateMode = UnixFileMode.DefaultFileOpenPermissions);

    // Set mode when creating directory
    public static DirectoryInfo CreateDirectory(string path, UnixFileMode unixCreateMode);
}

static partial class File
{
    public static SafeFileHandle OpenHandle (string path, System.IO.FileMode mode = System.IO.FileMode.Open, System.IO.FileAccess access = System.IO.FileAccess.Read, System.IO.FileShare share = System.IO.FileShare.Read, System.IO.FileOptions options = System.IO.FileOptions.None, long preallocationSize = 0);
    [EditorBrowsable(EditorBrowsableState.Never)]
    public static SafeFileHandle OpenHandle (string path, FileMode mode, FileAccess access, FileShare share, FileOptions options, long preallocationSize = 0); // Existing API, marked as Never Browsable to add a new default parameter.

    // Get/Set from SafeHandle, (.NET 7) https://github.com/dotnet/runtime/issues/20234
    public static UnixFileMode GetUnixFileMode(SafeFileHandle fileHandle);
    public static void SetUnixFileMode(SafeFileHandle fileHandle, UnixFileMode mode);
}

class FileSystemInfo
{
    // Get/Set from string path
    public UnixFileMode UnixFileMode { get; set; }
}

// Set mode when creating file, returns FileStream (by `File.Open`).
class FileStreamOptions
{
    // Set mode when creating file
    public UnixFileMode UnixCreateMode { get; set; } = UnixFileMode.DefaultFileOpenPermissions;
}

// `UnixFileMode` replaces `TarFileMode` (https://github.com/dotnet/runtime/issues/65951)
namespace System.Formats.Tar
{
-    enum TarFileMode { ... }
}
```

**Notes**

On Windows/non-Unix:

- The `unixCreateMode` arg on `Directory.CreateDirectory`/`File.OpenHandle` and `FileStreamOptions.UnixCreateMode` are ignored. These APIs can be used in in cross-platform code.

- `FileSystemInfo.UnixFileMode get` returns `UnixFileMode.None`.

- `FileSystemInfo.UnixFileMode set` throws PNSE.

APIs that create a file/directory (`Directory.CreateDirectory`, `FileStreamOptions.UnixCreateMode`, `File.OpenHandle`) have 'POSIX behavior':

- If the file/directory exists, the mode argument is ignored.

- The OS kernel filters the provided mode by the process `umask`. The umask protects users from unintentionally opening up permissions. A regular Linux user has a umask of `0002` which prevents it from giving write permissions to _other_. Use-cases that require the exact mode, must make an additional call to `SetUnixFileMode`. See example 3.

- _note_: creating and setting permissions is atomic to ensure permissions are applied immediately

`UnixFileMode` can replace `TarFileMode` introduced in https://github.com/dotnet/runtime/issues/65951.

`AccessPermissionMask`, `DefaultFileOpenPermissions` enum values.

- These are convenient masks. Example usage of these masks: https://github.com/dotnet/runtime/blob/b42adaddff9047e32c9d7f4145746dbb579758f2/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.Unix.cs#L45-L51 and https://github.com/dotnet/runtime/blob/b42adaddff9047e32c9d7f4145746dbb579758f2/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs#L167-L174 .

### API Usage

*Example 1*: limit a file to be read/written by the owner only:
```cs
File.SetUnixFileMode("filename", UnixFileMode.UserRead | UnixFileMode.UserWrite);
```

*Example 2*: exclude 'Other' access to entries in a directory:
```cs
var di = new DirectoryInfo("somedir");
foreach (var fi in di.GetFileSystemInfos())
{
  fi.UnixFleMode = fi.UnixFileMode & ~(UnixFileMode.OtherExecute | UnixFileMode.OtherRead | UnixFileMode.OtherWrite);
}
```

*Example 3*: create new files/directories while extracting a tar file with the exact permissions of the tar file.

```cs
if (entry.EntryType == EntryType.Directory)
{
    DirectoryInfo directoryInfo = Directory.CreateDirectory(path, entry.Mode); // filtered by umask
    di.UnixFileMode = entry.Mode;
}
else if (entry.EntryType == EntryType.RegularFile)
{
    using var file = File.OpenHandle(path, entry.Mode); // filtered by umask
    File.SetUnixFileMode(file, entry.Mode);
}
```

**Comment by tmds** [comment]
@eerhardt @carlossanlop @stephentoub @dotnet/area-system-io ptal.

**Comment by KalleOlaviNiemitalo** [comment]
File.OpenHandle with UnixFileMode looks useful for fixing the race condition in <https://github.com/AzureAD/microsoft-authentication-extensions-for-dotnet/pull/168> and could simplify [NuGetExtractionFileIO](https://github.com/NuGet/NuGet.Client/blob/5b0ee2562df6cf2c66cc9905a8a65f2b09ea5473/src/NuGet.Core/NuGet.Packaging/NuGetExtractionFileIO.cs).

**Comment by hez2010** [comment]
Should those API be put in a separate TFM `net7.0-linux` or `net7.0-unix` since they are not available on Windows platforms?

**Comment by eerhardt** [comment]
> Shouldn't those API be put in a separate TFM net7.0-linux or net7.0-unix since they are not available on Windows platforms?

Those TFMs don't exist. We don't have TFMs for `-linux` or `-unix`.

But instead we can use the SupportedOSPlatform and UnsupportedOSPlatform attributess that work with https://docs.microsoft.com/dotnet/standard/analyzers/platform-compat-analyzer. Then you will get `CA1416` warnings when you try using APIs that are unsupported in code that could run on Windows.

**Comment by eerhardt** [comment]
This looks great @tmds. Thanks for writing this up. I have the following questions:

1. Should we put the UnixFileMode on `File.Create` instead of `File.OpenHandle`?
2. In `Example 3`, why are we setting the UnixFileMode twice? Once on `File.OpenHandle` and then again with `File.SetUnixFileMode`?
3. >The mode argument is filtered by the process umask.
    - Do we need a way to set `umask` from .NET? Or is this not a typical usage scenario in an app?
4. Do we want convenience Flags values for UnixFileMode? For example, `UserExecute | UserWrite | UserRead` and `OtherExecute | OtherWrite | OtherRead`.

**Comment by eerhardt** [comment]
> >Do we need a way to set umask from .NET? Or is this not a typical usage scenario in an app?

> The umask is process wide, so it doesn't work well if several threads are setting it to a value they need.

I meant a public API that .NET developers could call to set the `umask`.

**Comment by tmds** [comment]
> I meant a public API that .NET developers could call to set the umask.

Because .NET is multi-threaded, it would end badly.

One thread doing 'untar', would set it to `umask` `0`.

While another thread doing `File.OpenWrite` would end up creating a file that is readable and writable by other.

**Comment by eerhardt** [comment]
> Because .NET is multi-threaded, it would end badly.

I can write multi-threaded C code. Does it end badly there as well?

> One thread doing 'untar', would set it to umask 0.

I'm not talking about `tar`. I'm talking about any .NET application. `umask` is a system-level API that allows applications to manipulate file modes. Do .NET applications need to control it?

**Comment by tmds** [comment]
I had a look. nodejs and python APIs that create files preserve the default behavior of applying umask.

Both also provide a umask function. Still, I don't think we should add it in .NET because of the multi-threading issues. The other APIs allow to set the permissions in a thread-safe way without using `umask`.

**Comment by tmds** [comment]
> Should we put the UnixFileMode on File.Create

Something like:

```cs
    public static FileStream Create(string path, UnixFileMode permissions, FileOptions options = FileOptions.None)
    	=> Create(path, permissions, DefaultBufferSize, options);
    public static FileStream Create(string path, UnixFileMode permissions, int bufferSize, FileOptions options = FileOptions.None)
    	=> new FileStream(OpenHandle(path, permissions, FileMode.Create, FileAccess.ReadWrite, FileShare.None, options), FileAccess.ReadWrite, bufferSize);
```

The `FileMode`/`FileAccess` match the existing `Create` but they are different from what I had picked for `OpenHandle`.

@eerhardt what do you think?

**Comment by KalleOlaviNiemitalo** [comment]
Setting umask in ProcessStartInfo would be safe, but if an application cares about that, then it likely wants to control ulimit and file descriptors as well, so just adding umask support there would not be worthwhile.

**Comment by jkotas** [comment]
>  UnixFileMode on File.Create

We have introduced `FileStreamOptions` that aggregate all file creation options to avoid the ever-growing list of overloads. Should the `UnixFileMode` be added on `FileStreamOptions` instead?

**Comment by tmds** [comment]
I've included it as a `FileStreamOption`, and extended the notes on the `umask` masking performed by the kernel.

**Comment by carlossanlop** [comment]
Thank you for proposing this. Getting UnixFileMode approved and implemented in 7.0 with all these usages would allow us to remove TarFileMode, which is what ended up being approved here (instead of UnixFileMode): https://github.com/dotnet/runtime/issues/65951

I think this proposal should explicitly mention that TarFileMode would be removed once UnixFileMode is approved. @bartonjs agree?

**Comment by eerhardt** [comment]
I think this proposal looks good enough to be marked `ready for review`. Thoughts @adamsitnik @Jozkee?

**Comment by eerhardt** [comment]
> Should we put the UnixFileMode on File.Create

Looking again, `File.Create` is really a convenience API that doesn't have all the possible options. For example, there is no `File.Create` that lets you specify a `FileShare`. Since setting the UnixFileMode is more of an advanced API, I don't think we need an overload to `File.Create` that takes a `UnixFileMode`. Devs can use one of the above proposed APIs `new FileStream` or `File.OpenHandle` if they need to create a file with a specific `UnixFileMode`.

**Comment by tmds** [comment]
> I think this proposal should explicitly mention that TarFileMode would be removed once UnixFileMode is approved.

I've added this to the proposal.

**Comment by tmds** [comment]
> I think this proposal looks good enough to be marked ready for review. Thoughts @adamsitnik @Jozkee?

@adamsitnik @Jozkee do you have feedback? Can this move to _ready for review_?

**Comment by tmds** [comment]
> Does the UnixFileMode sets atomically when the file or directory is created? If not, I think having only FileSystemInfo.UnixFileMode { get; set; } could suffice to address the same scenarios; but even if it sets atomically, it feels weird that the only overload to CreateDirectory takes a UnixFileMode but not any of the other properties already exposed.

Yes, it's atomic, and it is needed to restrict permissions immediately on creation.

> As you are already expanding FileStreamOptions, File.Open(string, FileStreamOptions) already covers file creation scenarios and you can still get the handle from the returning FileStream. I think this API is not needed.

I'll keep it in the proposal, and let the API reviewers kick it out.

> AccessPermissions = {User,Group,Other}{Execute,Read,Write}

This is a mask to filter out the access permissions. It could be left out. This is an example use-case:

https://github.com/dotnet/runtime/blob/b42adaddff9047e32c9d7f4145746dbb579758f2/src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarEntry.Unix.cs#L45-L51

I'll include some comments to the proposal based on your feedback.

**Comment by tmds** [comment]
@Jozkee @adamsitnik when it looks good to you, can you move the issue to _ready to review_?

**Comment by Jozkee** [comment]
Looks good as proposed, just a few more comments.

1. Expand `AccessPermissions` and DefaultFileOpenPermissions.
```cs
[System.FlagsAttribute]
public enum UnixFileMode
{
    AccessPermissions = UserRead | UserWrite | UserExecute | GroupRead | GroupWrite | GroupExecute | OtherRead | OtherWrite | OtherExecute,
    DefaultFileOpenPermissions = UserRead | UserWrite | GroupRead | GroupWrite | OtherRead | OtherWrite,
}
```

2. On `FileStreamOptions`, `InitialUnixFileMode` may be a better name than `UnixCreateMode`; we will also need to think if specifying this field should cause an exception or just be ignored when we are not creating a file.

3. Why use the word "Special" for `GroupSpecial` and `UserSpecial` values?

**Comment by KalleOlaviNiemitalo** [comment]
> we will also need to think if specifying this field should cause an exception or just be ignored when we are not creating a file

FileMode.Append, FileMode.Create, and FileMode.OpenOrCreate can either create a file or open an existing one. With these, specifying a UnixFileMode should not cause an exception, even if the call opens an existing file.
FileMode.Open and FileMode.Truncate can only open an existing file. With these, specifying a UnixFileMode is not useful and could be made to cause an exception.

However, if an application wants the effect of FileMode.OpenOrCreate but also needs to know whether the file was created or opened, I imagine it can:

1. Initialize FileStreamOptions with FileMode.CreateNew and some UnixFileMode.
2. Try to create the file with that.
3. Catch the IOException caused by an existing file.
4. Change FileStreamOptions.Mode to FileMode.Open, not changing the UnixFileMode in the FileStreamOptions.
5. Try to open the file with that.

If the UnixFileMode that the application set in step 1 causes an exception in step 5, and the application has to fix that by creating a new FileStreamOptions instance or by resetting to UnixFileMode.DefaultFileOpenPermissions, it will look somewhat pointless. I don't know how realistic this scenario would be, though.

**Comment by tmds** [comment]
> On FileStreamOptions, InitialUnixFileMode may be a better name than UnixCreateMode; we will also need to think if specifying this field should cause an exception or just be ignored when we are not creating a file.

The underlying `open` API only applies the mask when it creates the file. And when `open` is used with a flag that permits both creating and opening an existing file, the user has no way of distinguishing between these cases. The proposal suggests the .NET APIs to have the same semantics, which I described in the notes as:

```
If the file/directory exists, the mode argument is ignored.
```

> Why use the word "Special" for GroupSpecial and UserSpecial values?

These come from https://github.com/dotnet/runtime/issues/65951. Perhaps @carlossanlop can elaborate why the names contain `Special`? I included the alternative names `SetGroup`/`SetUser`. @carlossanlop wdyt about these alternatives?

> FileMode.Open and FileMode.Truncate can only open an existing file. With these, specifying a UnixFileMode is not useful and could be made to cause an exception.

The proposal is indeed to throw an exception in these cases:

```
If the FileMode never creates a file (it does not match: FileMode.*Create*), ArgumentException is thrown.
```

**Comment by KalleOlaviNiemitalo** [comment]
Throwing an exception for a suspicious FileMode is OK for the File.OpenHandle method in which UnixFileMode unixCreateMode is a required parameter. However, I think methods that take a FileStreamOptions parameter should ignore the UnixFileMode if FileStreamOptions.Mode does not allow creating a file.

**Comment by tmds** [comment]
> However, I think methods that take a FileStreamOptions parameter should ignore the UnixFileMode if FileStreamOptions.Mode does not allow creating a file.

That's right. Otherwise all `FileStreamOptions` that don't have a `*Create*` mode would stop working.

**Comment by adamsitnik** [comment]
Do we really need `GetUnixFileMode` and `SetUnixFileMode` for `Directory`, `File` and `FileSystemInfo`? Since it's not a cross-platform API I would expect that extending `FileSystemInfo` with the property should be enough.

```diff
static partial class Directory
{
-    public static UnixFileMode GetUnixFileMode(string path);
-    public static void SetUnixFileMode(string path, UnixFileMode mode);
}

static partial class File
{
-    public static UnixFileMode GetUnixFileMode(string path);
-    public static void SetUnixFileMode(string path, UnixFileMode mode);
}

class FileSystemInfo
{
    public UnixFileMode UnixFileMode { get; set; }
}
```

An alternative it to extend `Path` with two new methods (similar to `Path.Exists`)

> note: FileMode, FileAccess, FileShare have sensible defaults which are different from existing OpenHandle

This might confuse some of our users, moreover I am not sure which overload IDE like VS is going to suggest first. If the one that accepts `UnixFileMode` which is not a cross-platform concept and might throw on Windows then it's bad. I would rather not add `SafeFileHandle` methods in the first iteration.

```diff
static partial class File
{
-    public static SafeFileHandle OpenHandle(string path, UnixFileMode unixCreateMode, FileMode mode = FileMode.CreateNew, FileAccess access = FileAccess.Write, FileShare share = FileShare.None, FileOptions options = FileOptions.None, long preallocationSize = 0);

-    public static UnixFileMode GetUnixFileMode(SafeFileHandle fileHandle);
-    public static void SetUnixFileMode(SafeFileHandle fileHandle, UnixFileMode mode);
}
```

**Comment by tmds** [comment]
> Do we really need GetUnixFileMode and SetUnixFileMode for Directory, File and FileSystemInfo? Since it's not a cross-platform API I would expect that extending FileSystemInfo with the property should be enough.

I'll leave these out.

```cs
var mode = File.GetUnixFileMode(path);
```
becomes
```cs
var mode = new FileSystemInfo(path).UnixFileMode
```

> I am not sure which overload IDE like VS is going to suggest first.

I'll leave it out.

C# doesn't lend itself to extending an existing method with additional default arguments.

Without this overload, there is no way to create a `SafeFileHandle` without having to carry along a `FileStream` (for the `IDisposable`).

```cs
using SafeFileHandle handle = File.OpenHandle(path, unixCreateMode: mode);
```

becomes

```cs
using FileStream fs = new FileStream(path, new FileStreamOptions { UnixCreateMode = mode });
SafeFileHandle handle = fs.SafeFileHandle;
```

> I would rather not add SafeFileHandle methods in the first iteration.

Operating on a `string path` or a `SafeFileHandle` are functionally different. I'm going to keep these in the proposal because they match with the use-cases for https://github.com/dotnet/runtime/issues/20234.

API reviewers can still kick them out.

**Comment by bartonjs** [comment]
> C# doesn't lend itself to extending an existing method with additional default arguments.

There's a pattern for it, with the work on the library author.

* Add the signature of the new method with all of the defaulted parameters
* Remove the defaults from the existing member, and make it defer to the new, longer method.
* Mark the existing member `[EditorBrowsable(EditorBrowsableState.Never)]`

It doesn't work on virtuals, or for interfaces, but our guidelines say not to use defaulted parameters for either of those cases (except for a CancellationToken parameter)

Any existing compilations will call the old method, which just calls the new.  Any new compilations will prefer the new method, unless they happened to be specifying every argument in their code, then they recompile to exactly what they were.

**Comment by tmds** [comment]
@bartonjs thanks, I'll add back the `OpenHandle` overload using this pattern.

**Comment by carlossanlop** [comment]
> can I replace the Special* enum names with the Set* names?

I welcome any name that would make the most sense, @tmds.

**Comment by bartonjs** [comment]
Putting my API Review hat on a bit early... "But isn't it only SetUID/SetGID for (executable) files, and it means something different, like 'sticky' on directories?  I think that the POSIX header comments call it 'special', that seems like the more general name."

Just sharing :smile:

**Comment by KalleOlaviNiemitalo** [comment]
<https://man7.org/linux/man-pages/man0/sys_stat.h.0p.html> shows descriptions of those bits copied from POSIX, I believe. They don't involve "special".

The S_ISGID bit has been used for:

- On executable files, set the group ID of the process when executed.
- On files not executable by the group, replace fcntl()/lockf() advisory locking with [mandatory locking](https://www.kernel.org/doc/html/v5.14/filesystems/mandatory-locking.html).
- On directories, make new subdirectories inherit the group ID from this directory, not from the process.

(I wonder if the bit has been used for anything on block devices or character devices… like [biff](https://cgit.freebsd.org/src/tree/usr.bin/biff/biff.c?id=0fdf5bc531fbde4fa80a85614ba40230062ebde4) sets S_IXUSR and S_IXGRP on a tty.)

**Comment by tmds** [comment]
I'm going to replace the names in the proposal by `SetUser`/`SetGroup` because I think those are the most well-known names for these bits, and they represent the most common use-case.

API reviewers can make the final call.

**Comment by bartonjs** [comment]
Hm. Clearly some bad data got shoved into my head at some point.

New data accepted, position changed.  Rename away :smile:

**Comment by eerhardt** [comment]
> I'm going to replace the names in the proposal by SetUser/SetGroup because I think those are the most well-known names for these bits, and they represent the most common use-case.

It looks like golang calls them (roughly) the same: `ModeSetuid` and `ModeSetgid`.  https://pkg.go.dev/io/fs#FileMode

I don't see Java having names for these. https://docs.oracle.com/javase/8/docs/api/java/nio/file/attribute/PosixFilePermission.html. Looks like there are just `Read/Write/Execute` for `Group/Others/Owner`.

**Comment by terrajobst** [comment]
* Let's remove `Directory.OpenHandle`
    - We don't have an `OpenHandle` on Directory today
* `File`
    - Let's add overloads for `string path`
* Let's also remove `File.OpenHandle`
    - We don't have an overload today that allows setting Windows ACLs, it's a low level API
    - This would also have implications for if we use the 
* None of the APIs will be marked as Linux only as it seems we can and should make them work on Windows by setting WSL related metadata in NTFS
    - See https://docs.microsoft.com/en-us/windows/wsl/file-permissions
* `UnixFileMode`
    - Let's remove the mask because it messes with `ToString()`

```C#
namespace System.IO;

[System.FlagsAttribute]
public enum UnixFileMode
{
    None = 0,
    OtherExecute = 1,
    OtherWrite = 2,
    OtherRead = 4,
    GroupExecute = 8,
    GroupWrite = 16,
    GroupRead = 32,
    UserExecute = 64,
    UserWrite = 128,
    UserRead = 256,
    StickyBit = 512,
    SetGroup = 1024,
    SetUser = 2048
}
public partial class Directory
{
    public static DirectoryInfo CreateDirectory(string path, UnixFileMode unixCreateMode);
}
public partial class File
{
    public static UnixFileMode GetUnixFileMode(string path);
    public static UnixFileMode GetUnixFileMode(SafeFileHandle fileHandle);
    public static void SetUnixFileMode(string path, UnixFileMode mode);
    public static void SetUnixFileMode(SafeFileHandle fileHandle, UnixFileMode mode);
}
public partial class FileSystemInfo
{
    public UnixFileMode UnixFileMode { get; set; }
}
public partial class FileStreamOptions
{
    public UnixFileMode? UnixCreateMode { get; set; }
}
```

**Comment by stephentoub** [comment]
> public static DirectoryInfo CreateDirectory(string path, UnixFileMode unixCreateMode);

Do we need an equivalent for creating a file?  Otherwise, with the APIs I see approved above, it doesn't seem like there's a way for me to create a file and atomically have it have the "right" permissions, e.g. I might want a file locked down to just user, but I'll end up creating a file with the default permissions and by the time I've restricted them, someone else could have accessed it.

(Sorry if this was discussed; had to leave the meeting early...)

**Comment by Jozkee** [comment]
> Do we need an equivalent for creating a file? Otherwise, with the APIs I see approved above, it doesn't seem like there's a way for me to create a file and atomically have it have the "right" permissions

```cs
FileStreamOptions options = new()
{
    Mode = FileMode.Create,
    UnixCreateMode = UnixFileMode.UserRead | UnixFileMode.UserWrite
}

File.Open(filePath, options).Dispose();
```

**Comment by deeprobin** [comment]
Why is there no `OSPlatformAttribute` proposed? If I understand this right, this would only work on Unix-systems.

**Comment by tmds** [comment]
@Jozkee @adamsitnik @eerhardt you can assign this to me.

> Why is there no OSPlatformAttribute proposed? If I understand this right, this would only work on Unix-systems.

Let's see where it makes sense as part of the PR.

**Comment by stephentoub** [comment]
> Why is there no OSPlatformAttribute proposed?

"None of the APIs will be marked as Linux only as it seems we can and should make them work on Windows by setting WSL related metadata in NTFS"

**Comment by stephentoub** [comment]
> FileStreamOptions

Ah, yes, thanks.

**Comment by deeprobin** [comment]
> > Why is there no OSPlatformAttribute proposed?
> 
> "None of the APIs will be marked as Linux only as it seems we can and should make them work on Windows by setting WSL related metadata in NTFS"

Thanks. I overread that 🤦🏼‍♂️.

**Comment by teo-tsirpanis** [comment]
Fixed by #69980.

**Comment by webczat** [comment]
no chown/chgrp? anyway thanks for this one!
